### PR TITLE
perf(sync): Reduce db operations for accounts without CONDSTORE

### PR DESCRIPTION
**Pull request by JohannesGGE**
Fix: #6341

---

Queries (only related) with open mail app for 40min (approx. 170 mails in the mailbox)

before:
```
1367 	 primary UPDATE `oc_mail_messages` SET `flag_answered` = :flag_answered, `flag_deleted` = :flag_deleted, `flag_draft` = :flag_draft, `flag_flagged` = :flag_flagged, `flag_seen` = :flag_seen, `flag_forwarded` = :flag_forwarded, `flag_junk` = :flag_junk, `flag_notjunk` = :flag_notjunk, `flag_mdnsent` = :flag_mdnsent, `flag_important` = :flag_important, `updated_at` = :dcValue1 WHERE (`uid` = :uid) AND (`mailbox_id` = :mailbox_id)
```

after:
(So usually there are 2 queries per flag (138) (set true and set false). seen and notjunk are the only ones with doubled queries because for the other flags, either all mails or no mail have the flag in this specific mailbox.)
```
    138 	 primary UPDATE `oc_mail_messages` SET `flag_seen` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
    138 	 primary UPDATE `oc_mail_messages` SET `flag_notjunk` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_mdnsent` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_junk` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_important` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_forwarded` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_flagged` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_draft` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_deleted` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)
     69 	 primary UPDATE `oc_mail_messages` SET `flag_answered` = :dcValue1, `updated_at` = :dcValue2 WHERE (`uid` IN (:uids)) AND (`mailbox_id` = :dcValue3)

```

### CODE REVIEW

**Review by ChristophWurst — COMMENTED** [review]
The speedup is unparalleled

image




https://blackfire.io/profiles/compare/e2ee40f6-f178-401e-a403-f1f791907144/graph

:rocket:

File: lib/Db/MessageMapper.php [hunk 1]
```diff
@@ -373,24 +373,125 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
 			'partial sync ' . $account->getId() . ':' . $account->getName()
 		);
 
+		// MailboxId is the same for all messages according to updateBulk() call
+		$mailboxId = $messages[0]->getMailboxId();
+
+		$flags = array(
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:379]:
```suggestion
		$flags = array(
```

File: lib/Db/MessageMapper.php [hunk 2]
```diff
@@ -373,24 +373,125 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
...
+					$queryTrue->executeStatement();
+				}
+
+				$queryFalse = $this->db->getQueryBuilder();
+				$queryFalse->update($this->getTableName())
+					->set($flag, $queryFalse->createNamedParameter(0, IQueryBuilder::PARAM_INT))
+					->set('updated_at', $queryFalse->createNamedParameter($this->timeFactory->getTime(), IQueryBuilder::PARAM_INT))
+					->where($queryFalse->expr()->andX(
+						$queryFalse->expr()->in('uid', $queryFalse->createParameter('uids')),
+						$queryFalse->expr()->eq('mailbox_id', $queryFalse->createNamedParameter($mailboxId, IQueryBuilder::PARAM_INT))
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:486]: to experiment with: `AND flag=1`

to lower the number of accessed rows and hopefully the transaction size (for rollbacks and for replication)

Comment (ChristophWurst) [reply lib/Db/MessageMapper.php:486]: Seems worth pursuing 

image

Diff to the current HEAD

image

https://blackfire.io/profiles/compare/797f6164-4438-4c43-bcde-8d18dbd18b76/graph

Diff to main

image

https://blackfire.io/profiles/compare/eaa6c793-1f5b-4793-b13d-ea14a5b9ca20/graph

Comment (JohannesGGE) [reply lib/Db/MessageMapper.php]: :white_check_mark:

File: lib/Db/MessageMapper.php [hunk 3]
```diff
@@ -373,24 +373,127 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
 			'partial sync ' . $account->getId() . ':' . $account->getName()
 		);
 
+		// MailboxId is the same for all messages according to updateBulk() call
+		$mailboxId = $messages[0]->getMailboxId();
+
+		$flags = array(
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:379]:
```suggestion
		$flags = [
```

File: lib/Db/MessageMapper.php [hunk 4]
```diff
@@ -373,24 +373,127 @@ public function updateBulk(Account $account, bool $permflagsEnabled, Message ...
...
+			'flag_flagged',
+			'flag_seen',
+			'flag_forwarded',
+			'flag_junk',
+			'flag_notjunk',
+			'flag_mdnsent',
+			'flag_important',
+		);
+
+		$updateData = array();
```
Comment (ChristophWurst) [inline lib/Db/MessageMapper.php:392]:
```suggestion
		$updateData = [];
```

**Review by ChristophWurst — APPROVED** [review]
Please use modern array syntax

:rocket:

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**